### ISSUE TITLE
Cut and copy commands only work part of the time

### ISSUE THREAD

**Issue by eliddell** [pkulchenko/ZeroBraneStudio#759]
To reproduce:  In ZeroBrane 1.50, highlight some text with the mouse and invoke the cut or copy command using generic shortcut keys (Ctrl-X or Ctrl-C), the right-click context menu, or the main menu.  Do this two or three times.

Expected behaviour:  On every cut or copy, the text should end up on the system clipboard.  (It should also disappear from its original position if the cut command is used.  That part works.)

Actual behaviour:  On my system, the text _sometimes_ ends up on the clipboard.  About 50% of the time, it doesn't make it.  It can't be pasted back in, and doesn't show up on my clipboard viewer.  The program usually seems to get it right on odd-numbered attempts (first, third, fifth . . .)

ZeroBrane 1.30 had this problem as well.  It doesn't appear to be a bug in wxWidgets (unless it's peculiar to the Lua binding).

I'm on a Gentoo Linux system (64-bit).  I can give you more details on system library versions, etc. if you need them.

This has been driving me insane for quite a while now.

**Comment by pkulchenko** [comment]
@eliddell, can you also try in 1.60 (or the latest master, which should be the same version)? There have been several improvements to Unicode handling in copy/paste operations, which changed this logic in several aspect and I'm wondering if the issue is gone or if it's still present.

Are you copying regular text or does it include any Unicode characters (valid or invalid)?

**Comment by eliddell** [comment]
The problem appears to still be present in 1.60, unfortunately.  Very occasionally it will disappear for a while if I restart the IDE repeatedly.  I can't tell the difference between the "good" and "bad" restarts.

The text being copied and pasted is plain ASCII for the most part.  Some of the files I've been working with do contain a few UTF-8 extended characters, but it doesn't seem to make a difference either way, and the problem occurs even when all the files open are ASCII-only.  My system locale is en_CA.UTF-8 , for what it's worth.

One thing I should have mentioned and forgot about is that running the program from the command line always generates several messages of the following form:

(lua:[PID]): GLib-GObject-WARNING **: [path to where glib was compiled]/glib-2.50.3/gobject/gsignal.c:2641: instance '[hex address]' has no handler with id '[4-digit number]'

I don't know if that's related or not.

**Comment by pkulchenko** [comment]
@eliddell, thank you for the update. I don't have gentoo, but will see if I can reproduce on one of the Linux VMs I have; it may be related to your locale setting or to the Linux flavor as I haven't seen similar reports from other Linux users.

It doesn't seem like the warning you are seeing is related to this issue, although I can't be 100% certain about this.

**Comment by pkulchenko** [comment]
@eliddell, I haven't been able to reproduce this in Ubuntu, so I'll try to get a gentoo VM and see if I can reproduce it there...

**Comment by eliddell** [comment]
I'm not sure how effective that would be.  Gentoo systems are like snowflakes—no two alike—and my system was initially set up twelve years ago, so it has a lot of history and a lot of old program and settings cruft that would be difficult to duplicate.  Plus, there's no easy way to install Gentoo to a VM without following the entire installation process, and that can take a while.

Since you can't reproduce the problem on another distro, my guess is that this something in my particular system setup or USE-flag settings is confusing either your code or one of the libraries you're shipping.  libwx.so seems a likely culprit—maybe it doesn't like GTK2 being compiled with introspection disabled, or glib compiled with just about everything optional disabled.  Or it might not like my desktop environment (surviving KDE3 installs are not all that common these days), or some random distro-specific patch, or . . . well, you get the idea.

I think it might be better at this point for me to try to narrow down what's causing this myself—I've got a second Gentoo machine plus a couple of VMs I can tinker with.  I'll see if I can give it a shot this weekend.

**Comment by pkulchenko** [comment]
@eliddell, sounds good; thank you for the update.

One more thing to try is to rebuild the wxlua library and see if it changes the outcome. You can run the following command: `(cd build; ./build-linux.sh lua wxlua wxwidgets)`.

**Comment by eliddell** [comment]
Well, the answers I've gotten so far are . . . not very helpful.

Using a Gentoo VM with slightly outdated libraries and the LXDE desktop environment, the copy command seems to work correctly.

Using the same VM with the Trinity Desktop Environment (essentially the same as the KDE3 on my main machine), the copy command does *not* work properly.

Using a non-Gentoo LiveCD with Trinity (the specific LiveCD I picked was Devuan-based), the copy command works properly again.

So there's something going on that's specific to my combination of desktop environment and distro, but I still haven't figured out what.

I'm currently trying to rebuild wxlua inside the VM, but it looks like I'm going to have to rebuild git first, since I installed it with very minimal options that didn't include SSL support.  (By the way, you're no longer including the build directory in your releases—I had to grab the build-linux script manually from github.  Not a big deal, but I thought I'd mention it in case it was an oversight.)

If that doesn't work, I'll go fishing for a few more live media and see if I can reproduce this against a more common non-GTK desktop environment (which probably means KDE4/KDE5, ugh).  After that, I'm all out of ideas.  I can't even pack up the VM and send it to you, because everything on it is compiled specifically for an AMD CPU, which I doubt matches what you're using.

**Comment by pkulchenko** [comment]
> So there's something going on that's specific to my combination of desktop environment and distro, but I still haven't figured out what.

That's unfortunate; thank you for checking on these options.

> (By the way, you're no longer including the build directory in your releases—I had to grab the build-linux script manually from github. Not a big deal, but I thought I'd mention it in case it was an oversight.)

That's intentional; the packaged releases don't include `build/` and `t/` folders and some other files that are not needed for the users.

I'm interested, what *does* it copy when the operation fails? Does it keep the original value in the clipboard or is it emptied? Does it always work on the second try? I can probably add a workaround that does another copy before executing copy/cut, which may fix the issue (as a plugin, without changing the core functionality).

**Comment by eliddell** [comment]
When the operation fails, it does absolutely nothing that I can detect.  If the clipboard was empty, it remains empty; if it wasn't, the last item cut or copied stays on it.

Now for the good news:  recompiling wxlua according to your directions seems to have fixed the copy command.  So (if I'm correctly understanding what is and isn't included in your default package), the bug is probably due to my system copy of wxWidgets not matching up with the one you used when creating the copy of wxlua you're shipping.  I still have no idea how LXDE was able to get around the problem, though.

Fixing it properly (or at least fixing it so that I don't have to do an unnecessary wxWidgets recompile every time I update ZeroBrane) is probably going to require my packaging wxlua for Gentoo so that I can replace the bundled dependency.  That's going to be . . . interesting, but I think it's outside the scope of anything you need to worry about.  Regardless, I don't think building a specialized plugin is the way to approach it when there's a better fix available (and when it affects so very few users).

**Comment by pkulchenko** [comment]
@eliddell, thank you for the update. Now onto the details.

wxlua library included with the IDE actually doesn't use your system copy of wxwidgets as it only provides a thin wrapper around wxwidgets, so is compiled together with it and is bundled with the IDE. In your case, `clibs/libwx.so` includes wxlua along with wxwidgets 3.1.x code.

Unfortunately you may not be able to package the correct version of wxlua for Gentoo as I'm using the version of wxlua that I updated to be compatible with wxwidgets 3.1, but these changes have never been merged upstream, so if the "official" version of wxlua is used, you won't get any new functionality added with wxwidgets 3.1 upgrade (as this is when I switched to using my own fork of wxlua). While it may continue to work (as I try to maintain compatibility with wxwidgets going back to 2.9.5), some of the checks may break as when checking for 3.1, the code assumes some new API functions to be available, but they won't be, because the official wxlua doesn't include them.

Long story short, custom plugin may not be such a bad option (assuming it fixes the issue) ;).

My wxlua changes are all available here: https://github.com/pkulchenko/wxlua/commits/wxwidgets311 (you can do a diff with the master if you want to see the commits I added). I usually merge the changes from the master branch and rebase my changes on top.

**Comment by eliddell** [comment]
Recompiling wxlua fixed the issue on my main system too, so a plugin shouldn't be necessary.

As for packaging wxlua with your patchset, it actually wouldn't be all that difficult to write what's known as a "live ebuild" for it (essentially a package that pulls directly from your source control).  However, it looks like the last version of wxGTK in the main Gentoo repository at the moment is 3.0.2.0, and I'm not going to bump that too, since the maintainer has made it unusually complicated to do so.  I may revisit this when 3.1 becomes available.

**Comment by pkulchenko** [comment]
@eliddell, thank you for the update. I'll close the ticket then, since you have a workaround. Let me know if you run into any issues in rebuilding or packaging my branch.
